### PR TITLE
ILCompiler.pkgproj: remove 'runtime.json' when DotNetBuildSourceOnly.

### PR DESCRIPTION
**Pull request by tmds**
Contributes to https://github.com/dotnet/source-build/issues/1215

@dsplaisted @baronfel @ViktorHofer @jkotas ptal

cc @omajid

### CODE REVIEW

File: src/installer/pkg/projects/Microsoft.DotNet.ILCompiler/Microsoft.DotNet.ILCompiler.pkgproj [hunk 1]
```diff
@@ -4,6 +4,20 @@
...
+         references the corresponding 'runtime.<rid>.Microsoft.DotNet.ILCompiler' package for restore.
+
+         With a source-built SDK, we support bundling the source-built 'runtime.<rid>.Microsoft.DotNet.ILCompiler'
+         package as a pack. Since packs can not be restored by NuGet, the 'runtime.json' reference for the
+         source-built rid leads to NuGet restore errors.
+
+         To work around these errors we remove the 'runtime.json' file completely on the source-built SDK.
+         'ProcessFrameworkReferences' is responsible for either picking up the bundled pack (when targetting the source-built rid)
+         or to add a 'runtime.<rid>.Microsoft.DotNet.ILCompiler' package for download. -->
+    <IncludeRuntimeJson Condition="'$(DotNetBuildSourceOnly)' == 'true'">false</IncludeRuntimeJson>
```
Comment (tmds) [inline src/installer/pkg/projects/Microsoft.DotNet.ILCompiler/Microsoft.DotNet.ILCompiler.pkgproj:18]: This depends on https://github.com/dotnet/arcade/pull/14805.

**Comment by ViktorHofer** [comment]
@ericstj can you please take a look? (This is about pkgproj infra)

**Comment by ericstj** [comment]
Can we just remove this completely?  We shouldn't be relying on runtime.json anywhere anymore.  We removed that from the product in 2.0.  Anything that needs support like this should have an SDK doing the package selection.

**Comment by dsplaisted** [comment]
> Can we just remove this completely? We shouldn't be relying on runtime.json anywhere anymore. We removed that from the product in 2.0. Anything that needs support like this should have an SDK doing the package selection.

Per conversation in https://github.com/dotnet/source-build/issues/1215, it sounds like using a `PackageReference` to `Microsoft.DotNet.ILCompiler` is still supported, and for that scenario to work the runtime.json needs to be there to acquire the RID-specific ILCompiler package.

**Comment by jkotas** [comment]
> > Can we just remove this completely? We shouldn't be relying on runtime.json anywhere anymore. We removed that from the product in 2.0. Anything that needs support like this should have an SDK doing the package selection.

> Per conversation in https://github.com/dotnet/source-build/issues/1215, it sounds like using a PackageReference to Microsoft.DotNet.ILCompiler is still supported, and for that scenario to work the runtime.json needs to be there to acquire the RID-specific ILCompiler package.

We need a simple way to override the default version of `Microsoft.DotNet.ILCompiler`. It is frequently used to trouble shoot issues and to provide workarounds for customers. Microsoft.DotNet.ILCompiler package reference works perfectly for this purpose today.

Can we make it work equally well without runtime.json?

**Comment by dsplaisted** [comment]
> We need a simple way to override the default version of `Microsoft.DotNet.ILCompiler`. It is frequently used to trouble shoot issues and to provide workarounds for customers. Microsoft.DotNet.ILCompiler package reference works perfectly for this purpose today.
> 
> Can we make it work equally well without runtime.json?

Rather than a `PackageReference`, I think you could do this:

```
<ItemGroup>
    <KnownILCompilerPack Update="Microsoft.DotNet.ILCompiler" ILCompilerPackVersion="8.0.5"/>
</ItemGroup>
```

I think this would work with an ILCompiler package without the runtime.json and the logic we already have in the SDK today.

**Comment by jkotas** [comment]
I dislike that we would be moving from PackageReference to SDK magic. I think PackageReferences are much cleaner and easier to understand from user point of view.

If we agree that the SDK magic is right way to go, I guess it can work. @agocke @MichalStrehovsky Do you see any problems with it? We would have to update the docs and chase down all places that depend on the current behavior. It may be best to do it separately from enabling source build for native AOT toolchain.

> We shouldn't be relying on runtime.json anywhere anymore. We removed that from the product in 2.0. Anything that needs support like this should have an SDK doing the package selection.

`runtime.json` may not be officially documented, but it is the only way we have to provide good user experience for nuget packages with (large) platform-specific binaries. Both our and 3rd party packages depend on it for this reason.

**Comment by MichalStrehovsky** [comment]
> If we agree that the SDK magic is right way to go, I guess it can work. @agocke @MichalStrehovsky Do you see any problems with it? We would have to update the docs and chase down all places that depend on the current behavior. It may be best to do it separately from enabling source build for native AOT toolchain.

I don't see an issue other than what Jan already wrote. If we do this change, we'll need to consider KnownILCompilerPack and ILCompilerPackVersion basically a "public API" so that we don't have to redo chasing down all places that depend on this or mention this next time it changes. If we're doing a break, it would be ideal if this mechanism is standardized with other tools that ship in NuGets - ILLink and crossgen off the top of my head, so that we know what we have is going to last.

@sbomer is there an existing mechanism when a user needs to pick up a new/specific version of ILLink if they need a different one from the one that ships with the SDK?

**Comment by jkotas** [comment]
> it would be ideal if this mechanism is standardized with other tools that ship in NuGets - ILLink and crossgen off the top of my head

There are number of other tools. For example, you can reference a specific Roslyn version too - via a package reference. I think the only meaningful standard are package refences. Everything else are bespoke solutions that won't ever be consistent across the ecosystem.

**Review by jkotas — APPROVED** [review]

**Review by ViktorHofer — APPROVED** [review]

**Pull request merged**